cmd/skopeo/registry_utils.go

```diff
@@ -0,0 +1,303 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"sort"
+	"strings"
+	"sync"
+	"text/tabwriter"
+)
+
+// Registry stores information about the targeted docker registry to
+// to optimize the numerous calls to registry API
+type Registry struct {
+	baseURL      string
+	catalogURL   string
+	authType     string
+	regUsername  string
+	regPassword  string
+	repos        []string
+	catalogToken string
+}
+
+const (
+	tokenAuth string = "token"
+	basicAuth string = "basic"
+	noneAuth  string = "none"
+)
+
+// NewRegistry creates a new Registry struct with parsed/sanitized URLs
+func NewRegistry(rawURL string) (Registry, error) {
+	parsedURL, err := url.Parse(rawURL)
+	if err != nil {
+		return Registry{}, fmt.Errorf("Failed to parse Registry URL: %v", err)
+	}
+	baseURL := parsedURL.String()
+	parsedURL.Path = path.Join(parsedURL.Path, "v2/_catalog")
+	catalogURL := parsedURL.String()
+
+	return Registry{baseURL: baseURL, catalogURL: catalogURL}, nil
+}
+
+// Determines auth type to optimize later registry calls
+func (r *Registry) getAuthType() error {
+	resp, err := http.Get(r.catalogURL)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to Catalog URL: %v", err)
+	}
+
+	headerAuth := resp.Header.Get("www-authenticate")
+	headerAuthLower := strings.ToLower(headerAuth)
+
+	if strings.Count(headerAuthLower, "bearer") > 0 {
+		fmt.Println("Using Token Auth")
+		r.authType = tokenAuth
+	} else if strings.Count(headerAuthLower, "basic") > 0 {
+		fmt.Println("Using Basic Auth")
+		r.authType = basicAuth
+	} else {
+		fmt.Println("Using No Auth")
+		r.authType = noneAuth
+	}
+	return nil
+}
+
+// Uses r.authType to abstract Authentication
+func (r *Registry) applyAuth(req *http.Request) error {
+	if r.authType == tokenAuth {
+		var token string
+		// Each catalog request returns a limit of 100 repos at a time.
+		// This conditional saves a few requests for large registry catalogs.
+		// By resetting r.catalogToken at end of getV2Catalog(), this
+		// can also be used to handle auth for getting tags.
+		// This could could be improved.
+		if len(r.catalogToken) > 0 {
+			token = r.catalogToken
+		} else {
+			// err must be declared to prevent token from being re-declared with ":=" below
+			var err error
+			token, err = getToken(req.URL.String())
+			if err != nil {
+				return err
+			}
+			r.catalogToken = token
+		}
+		req.Header.Set("Authorization", "Bearer "+token)
+	} else if r.authType == basicAuth {
+		req.SetBasicAuth(r.regUsername, r.regPassword)
+	} else {
+		// Do Nothing
+	}
+	return nil
+}
+
+// Prints list of registry repos with all available tags
+func (r *Registry) getAllReposWithTags() error {
+	err := r.getAuthType()
+	if err != nil {
+		return err
+	}
+	r.getV2Catalog()
```

Comment:
The full catalog may not be available at all, or it may return only a few entries. So, careful… I suppose it’s fine to expose this functionality if the registry does provide it, but we need to be be very careful in documenting this.